### PR TITLE
Turn DMF properties into DMFProperty types

### LINKED ISSUE #1730: Goon: Say messages which contain spaces don't send
example: press T, type "test", press OK - message sends
bad example: press T, type "test test", press OK - message does not send

### PR DESCRIPTION
**Pull request by amylizzle**
`winget` stuff is more complicated than OD can handle right now, and it needs some type information needs to be passed along to process it properly. Notably pos and size vars can be accessed by their individual components, or formatted as json like `{"x":123, "y":321}`. Also pretty much all of the `as blah` features have different behaviour depending on the type of the var.

Requires: #1789 
This PR implements:
- [x] Type information for DMF properties
- [x] Per-type serialization for each type of property
- [x] Fully implements the `as escaped/json/etc.` `winget` commands (fixes #1730) 
- [x] Standardises `TryGetProperty()` and default values for descriptors

I'm not doing this:
- Allows `thing.pos.x` `winget` access
because nobody uses it and it'd be a huge pain. I guess we'll do it later if it's needed.

### CODE REVIEW

**Comment by amylizzle** [comment]
This is now finished, in that I'm not intending to add anymore to it. However, our positioning code is kinda fucky (see: the buttons in the winget test window), and that probably ought to be fixed in a separate PR because this is already huge.

Also I'm not doing the .x/.y components of size/pos DMF props, because it'd be another rewrite of winget processing and ceebs.

**Comment by amylizzle** [comment]
All required prs are merged, this is now good to go after review

File: OpenDreamClient/Interface/DMF/DMFProperty.cs [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+    public DMFPropertyVec2(int x, int y) {
+        X = x;
+        Y = y;
+    }
+
+    public DMFPropertyVec2(string value) {
+        if(value.Equals("none",StringComparison.InvariantCultureIgnoreCase)){
+            X = 0;
+            Y = 0;
+        }
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DMF/DMFProperty.cs:146]:
```suggestion
            Y = 0;
            return;
        }
```

File: OpenDreamClient/Interface/DMF/DMFProperty.cs [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+
+        X = int.Parse(parts[0]);
+        Y = int.Parse(parts[1]);
+    }
+
+    public DMFPropertyVec2(Vector2 value) {
+        X = (int)value.X;
+        Y = (int)value.Y;
+    }
+
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DMF/DMFProperty.cs:158]:
```suggestion
```

File: OpenDreamClient/Interface/Descriptors/ControlDescriptors.cs [hunk 3]
```diff
@@ -1,54 +1,101 @@
...
+    public DMFPropertyBool RightClick = new(false);
+    [DataField("font-family")]
+    public DMFPropertyString FontFamily = new("");
+    [DataField("font-size")]
+    public DMFPropertyNum FontSize = new(0);
+    [DataField("font-style")]
+    public DMFPropertyString FontStyle = new("");
+    [DataField("on-size")]
+    public DMFPropertyString OnSize = new("");
+
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/Descriptors/ControlDescriptors.cs:51]:
```suggestion
```

File: OpenDreamClient/Interface/Descriptors/ControlDescriptors.cs [hunk 4]
```diff
@@ -1,54 +1,101 @@
...
+    [DataField("is-minimized")]
+    public DMFPropertyBool IsMinimized = new(false);
+    [DataField("is-maximized")]
+    public DMFPropertyBool IsMaximized = new(false);
+    [DataField("alpha")]
+    public DMFPropertyNum Alpha = new(255);
+    [DataField("statusbar")]
+    public DMFPropertyBool StatusBar = new(false);
+    [DataField("transparent-color")]
+    public DMFPropertyColor? TransparentColor = null;
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/Descriptors/ControlDescriptors.cs:68]: Should this be non-nullable and default to transparent?

File: OpenDreamClient/Interface/Descriptors/ControlDescriptors.cs [hunk 5]
```diff
@@ -99,132 +146,224 @@ public sealed partial class WindowDescriptor : ControlDescriptor {
...
+    [DataField("max-lines")]
+    public DMFPropertyNum MaxLines = new(1000);
+    [DataField("link-color")]
+    public DMFPropertyColor LinkColor = new(Color.Blue);
+    [DataField("visited-color")]
+    public DMFPropertyColor VisitedColor = new(Color.Purple);
+    [DataField("image")]
+    public DMFPropertyString Image = new("");
+    [DataField("enable-http-images")]
+    public DMFPropertyBool EnableHTTPImages = new(false);
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/Descriptors/ControlDescriptors.cs:224]: Rename to `EnableHttpImages`

File: OpenDreamClient/Interface/Descriptors/ControlDescriptors.cs [hunk 6]
```diff
@@ -99,132 +146,224 @@ public sealed partial class WindowDescriptor : ControlDescriptor {
...
-    public int IconSize;
+    public DMFPropertyString OnHideCommand = new("");
+
 }
 
 public sealed partial class ControlDescriptorBrowser : ControlDescriptor {
+    [DataField("show-history")]
+    public DMFPropertyBool ShowHistory = new(false);
+    [DataField("show-url")]
+    public DMFPropertyBool ShowURL = new(false);
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/Descriptors/ControlDescriptors.cs:281]: Rename to `ShowUrl`

File: OpenDreamClient/Interface/Descriptors/ControlDescriptors.cs [hunk 7]
```diff
@@ -99,132 +146,224 @@ public sealed partial class WindowDescriptor : ControlDescriptor {
...
+    [DataField("visited-color")]
+    public DMFPropertyColor VisitedCOlor = new(Color.Purple);
+    [DataField("current-cell")]
+    public DMFPropertyVec2 CurrentCell = new(0,0);
+    [DataField("show-names")]
+    public DMFPropertyBool ShowNames = new(true);
+    [DataField("small-icons")]
+    public DMFPropertyBool SmallIcons = new(false);
+    [DataField("enable-http-images")]
+    public DMFPropertyBool EnableHTTPImages = new(false);
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/Descriptors/ControlDescriptors.cs:333]: Rename to `EnableHttpImages`

File: TestGame/TestInterface.dmf [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	elem "testwindow"
+		type = MAIN
+		size = 200x100
+		title = "popup"
+		is-visible = false
+	elem "testwindowlabel"
+		type = LABEL
+		text = "I am a test"
+
+window "wingetwindow"
```
Comment (wixoaGit) [inline TestGame/TestInterface.dmf:149]: I don't like having this extra window show up when using TestGame

File: TestGame/TestInterface.dmf [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+		type = INPUT
+		pos = 5,0
+		size = 190,19
+		text = "this is some test text"
+	elem "wingetbuttonarg"
+		type = BUTTON
+		pos = 0,40
+		size = 0,20
+		text = "as arg"
+		command = "wingettextverb \"arg: [[wingetwindow.background-color as arg]]\""
```
Comment (wixoaGit) [inline TestGame/TestInterface.dmf:165]: Think you meant to keep this as `wingetinput.text`

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Comment by wixoaGit** [comment]
TG contains this in its DMF:
`font-size = 6pt`

The number property doesn't recognize this and throws an exception, leading to the interface never loading.

File: OpenDreamClient/Interface/Controls/InterfaceControl.cs [hunk 10]
```diff
@@ -8,17 +8,16 @@ namespace OpenDreamClient.Interface.Controls;
...
+    public DMFPropertyVec2 Size => ControlDescriptor.Size;
+    public DMFPropertyVec2 Pos => ControlDescriptor.Pos;
+    public DMFPropertyVec2? Anchor1 => ControlDescriptor.Anchor1;
+    public DMFPropertyVec2? Anchor2 => ControlDescriptor.Anchor2;
 
     protected ControlDescriptor ControlDescriptor => (ControlDescriptor) ElementDescriptor;
 
     private readonly ControlWindow _window;
 
-    [SuppressMessage("ReSharper", "VirtualMemberCallInConstructor")]
```
[the remark this answers is not in the document]
Comment (amylizzle) [reply OpenDreamClient/Interface/Controls/InterfaceControl.cs:21]: I wasn't getting the warning, but I'll put it back.

File: TestGame/TestInterface.dmf (hunk 8 above)
Comment (amylizzle) [reply TestGame/TestInterface.dmf:149]: I'll add another toggle

**Comment by amylizzle** [comment]
> TG contains this in its DMF: `font-size = 6pt`
> 
> The number property doesn't recognize this and throws an exception, leading to the interface never loading.

This is actually an error in TG's DMF - it should really be `font-size = 6`, but I'll make the parsing more permissive and maybe throw a warning instead.

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**